zilliqa/src/api/zil.rs

```diff
@@ -1209,56 +1270,32 @@ fn get_smart_contract_state_internal(
         unreachable!()
     };
 
-    let is_scilla = account.code.scilla_code_and_init_data().is_some();
-    if is_scilla {
-        let limit = node.config.state_rpc_limit;
-
+    if account.code.scilla_code_and_init_data().is_some() {
         let trie = state.get_account_trie(address)?;
-        for (i, (k, v)) in trie.iter().enumerate() {
-            if i >= limit {
-                return Err(anyhow!(
-                    "State of contract returned has size greater than the allowed maximum"
-                ));
-            }
 
-            let (var_name, indices) = split_storage_key(&k)?;
-            if let Some(ref x) = requested_varname {
-                if *x != var_name {
-                    continue;
-                }
-            }
-            let mut var = result.entry(var_name.clone());
-
-            for index in indices {
-                if let Some(ref x) = requested_indices {
-                    if !x.contains(&index.to_hex()) {
-                        continue;
-                    }
-                }
-                let next = var.or_insert_with(|| Value::Object(Default::default()));
-                let Value::Object(next) = next else {
-                    unreachable!()
-                };
-                let key: String = serde_json::from_slice(&index)?;
-                var = next.entry(key.clone());
-            }
-
-            var.or_insert(serde_json::from_slice(&v)?);
+        let indicies_encoded = requested_indices
+            .iter()
+            .map(|x| serde_json::to_vec(&x))
+            .collect::<std::result::Result<Vec<_>, _>>()?;
+        let k = storage_key(&var_name, &indicies_encoded);
+        let v = match trie.get(&k)? {
```

Comment:
I don't recall the intended behaviour of this API but do we need to handle the case where the user passes a requested index as a prefix and expects us to return many keys with that prefix? @chetan-zilliqa might be able to advise on some example requests and responses from ZQ1.

Reply:
Yeah if there's some examples to give more info on what's the desired behaviour here that'd be good because the documentation isn't super detailed for this one so I was filling in the blanks a little. I was also guessing that we needed `_balance` returned from this too.

Reply:
Fetching a complete contract state is cumbersome and bandwidth intensive. If a user is only interested in substate(for eg: perticular contract variable or key) then it can be fetched from the contract.
Eg:
```
curl --request POST \
  --url https://api.devnet.zilliqa.com/ \
  --header 'Content-Type: application/json' \
  --data '{
    "id": "1",
    "jsonrpc": "2.0",
    "method": "GetSmartContractSubState",
	"params": ["zil15tjx2l0gzzxaxucwk50stgw5smthhckl8qcqmf","bnum_req", []]
}'

curl --request POST \
  --url https://api.devnet.zilliqa.com/ \
  --header 'Content-Type: application/json' \
  --data '{
    "id": "1",
    "jsonrpc": "2.0",
    "method": "GetSmartContractSubState",
	"params": ["zil15lr86jwg937urdeayvtypvhy6pnp6d7p8n5z09","ssnlist", []]
}'

curl --request POST \
  --url https://api.devnet.zilliqa.com/ \
  --header 'Content-Type: application/json' \
  --data '{
    "id": "1",
    "jsonrpc": "2.0",
    "method": "GetSmartContractSubState",
	"params": ["0xa7c67d49c82c7dc1b73d231640b2e4d0661d37c1","deposit_amt_deleg", ["0xec31349b7f17ea28ff59c7586c7734be23e43443"]]
}'
curl --request POST \
  --url https://api.devnet.zilliqa.com/ \
  --header 'Content-Type: application/json' \
  --data '{
    "id": "1",
    "jsonrpc": "2.0",
    "method": "GetSmartContractSubState",
	"params": ["0xa7c67d49c82c7dc1b73d231640b2e4d0661d37c1","balances", ["0x406111ecfd389ad6e8a38536d72689dd5010b11a"]]
}'

curl --request POST \
  --url https://api.devnet.zilliqa.com/ \
  --header 'Content-Type: application/json' \
  --data '{
    "id": "1",
    "jsonrpc": "2.0",
    "method": "GetSmartContractSubState",
	"params": ["a2e4657de8108dd3730eb51f05a1d486d77be2df","stake_ssn_per_cycle", []]
}'


```

Reply:
Above is the example is of staking contract  - https://viewblock.io/zilliqa/address/zil15lr86jwg937urdeayvtypvhy6pnp6d7p8n5z09?tab=state

Reply:
@maxconway @JamesHinshelwood

Reply:
So I've tried all those, they all behave as I'd expect I think, what's your conclusion, @chetan-zilliqa? I've had another look at the trie code and I'm guessing what you're getting at is using iter_by_prefix rather than get, James? I'm not sure how constructing the prefix will interact with de-jsonifying the indices, but I'd imagine I'll be able to work it out.

Reply:
Right, @JamesHinshelwood the prefix handling capability is added